Gforth version

lonetech opened the pull request:
Forth is a stack based, minimal syntax, extremely imperative programming language which encourages tacit code. It's one of the early languages, and the source of phrases such as "refactor". While there exist standards, many Forths do things subtly differently, and programs frequently use shortcut words. Forth's design makes it very easy to bootstrap, and it is therefore popular for small (e.g. sectorforth) or exotic (e.g. pbforth) targets.

This is a reasonably complete implementation of bye bye hello world in [Gforth](https://gforth.org/). The main quirk is some assumption the user won't enter very large numbers, and the countdown loop uses -do -loop which is not standard.

Commit:
Gforth version

Commit:
Smarter (simpler) number parsing

Don't parse negative numbers, and >number is standard too.

c4augustus commented:
So far there is only a source file bbhw.fs and no run script with an automatic build or a least notice of where to download an installable binary, as the J and Dyalog APL run scripts provide. Also, looking at the source it seems to provide prompts that do not match the spec of the other languages, " Count? " instead of "countdown: ", and the error message also does not match.

c4augustus reviewed (changes requested):
So far there is only a source file bbhw.fs and no run script with an automatic build or a least notice of where to download an installable binary, as the J and Dyalog APL run scripts provide. Also, looking at the source it seems to provide prompts that do not match the spec of the other languages, " Count? " instead of "countdown: ", and the error message also does not match.

Commit:
gforth: Adjusted messages to match other versions

Commit:
gforth: avoid "holds" for compatibility with 0.7.3

Commit:
Added get and run scripts for gforth

Commit:
gforth: Quote bad input and check for empty input

c4augustus reviewed (changes requested):
gforth doesn't build on macOS using clang 15.0.0: "/Users/humano/z/v/d/es/plb/bbhw/code/forth/gforth/gforth-0.7.3/lib/gforth/0.7.3/libcc-named/libffi.c:95:9: error: call to undeclared function 'ffi_prep_closure'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
  sp[2]=ffi_prep_closure1((void *)(sp[2]),(void *)(sp[1]),(void *)(sp[0]));"

And building it on WSL 1 (Ubuntu 18.04) crashes my WSL and Windows (fork bomb?) using GCC Ubuntu 7.5.0.

Building GNU software is so much fun.

Commit:
pforth version

lonetech commented:
Well, obviously I can't test on your systems. 
gforth is already packaged in Debian (and the derivative Ubuntu), and homebrew (which is more likely to have whatever macosx requires to work).
Crashing the OS by running this should be completely impossible (but I have never attempted WSL).
Also I didn't request any parallel build, gforth took 13s.
It appears macosx specific for gforth to require libffi, and presumably they changed what the compiler accepts in the ten years since gforth 0.7.3. The brew formula also specifies emacs and libtool. 

I could also port to other forths such as pforth, yforth, zForth, 4th, wForth, RetroForth, durexforth, pbforth, ...

samirose commented on code/forth/gforth/get:
```diff
@@ -0,0 +1,11 @@
+#! /bin/sh -e
+# https://gforth.org/
+VER=0.7.3
+PREFIX="$PWD/gforth"
+wget -c https://ftp.gnu.org/gnu/gforth/gforth-$VER.tar.gz
+tar xzf gforth-$VER.tar.gz
+cd gforth-$VER
+./configure --prefix="$PREFIX"
+make
+make install
+
```
I managed to build gforth locally and run the bbhw.fs with it on my macOS (Sonoma 14.5) by building also the gforth dependencies libffi and libtool, see the suggestion below.

```suggestion
GFORTH_VER=0.7.3
PREFIX="$PWD/gforth"

 # libffi: gforth dependency
wget -c https://github.com/libffi/libffi/releases/download/v3.4.6/libffi-3.4.6.tar.gz
tar xzf libffi-3.4.6.tar.gz
pushd libffi-3.4.6
./configure --prefix="$PREFIX" --disable-debug --disable-docs
make -j
make install
popd

# libtool: gforth dependency
wget -c https://ftpmirror.gnu.org/libtool/libtool-2.4.7.tar.gz
tar xzf libtool-2.4.7.tar.gz
pushd libtool-2.4.7
./configure --prefix="$PREFIX"
make -j
make install
popd

# gforth
wget -c https://ftp.gnu.org/gnu/gforth/gforth-$GFORTH_VER.tar.gz
tar xzf gforth-$GFORTH_VER.tar.gz
pushd gforth-$GFORTH_VER
LTDL_LIBRARY_PATH="$PREFIX/lib" ./configure --prefix="$PREFIX"
make -j
make install
popd
```

One possible caveat: I use Homebrew and I have gforth and all its dependencies installed using it. The installed versions seem to mess somewhat with the gforth configure script, which e.g. detects the installed libtool as glibtool and even that I have gforth installed.

c4augustus replied:
gforth is working on WSL 1 (Ubuntu 18.04) except for the which -s argument in the run script. pforth run script is not passing command line args, but the bbhw works otherwise as normal. Please fix these two issues and I will approve this PR. As to macOS I will just assume that anyone installing gforth via homebrew or MacPorts will be fine (I no longer use neither homebrew nor MacPorts because they pollute the global environment).

samirose replied:
The caveat I mention above is real: gforth seems to require gforth for bootstrapping its own build. All instructions for building it I can find point to obtaining a binary somewhere and making it accessible to the gforth configure script. Also libffi is probably not needed as the Homebrew-installed gforth seems to link to the system library `usr/lib/libffi.dylib`.
```shell
% otool -L $(which gforth)
/opt/homebrew/bin/gforth:
	/usr/lib/libffi.dylib (compatibility version 1.0.0, current version 30.0.0)
	/opt/homebrew/opt/libtool/lib/libltdl.7.dylib (compatibility version 11.0.0, current version 11.2.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1319.0.0)
```

samirose replied:
> As to macOS I will just assume that anyone installing gforth via homebrew or MacPorts will be fine (I no longer use neither homebrew nor MacPorts because they pollute the global environment).

I can confirm that the bbhw.fs works with Homebrew-installed gforth 0.7.3 on macOS.

lonetech replied:
The bootstrapping is for kernl{16,32,64}[lb].fi as far as I can tell. They are included with the release, but also rebuilt from source (kernel/main.fs) in the normal make process. A separate gforth is not required. 

pforth actually doesn't support command line arguments.

lonetech committed:
Update code/forth/gforth/get

get script extended with more dependencies by samirose

Co-authored-by: Sami Rosendahl <sami.rosendahl@gmail.com>

Commit:
gforth get/run scripts bash adjustments

type as alternative to which, quote variables for caution

Commit:
pforth fixes

Drop count for 0 loop
Read entire input line for bbhw word
Make run script explain command line arguments are not supported

c4augustus commented:
One more change. Both pforth and gforth give the message "Invalid input" when it should be "Invalid countdown".

lonetech committed:
Update bbhw.fs

lonetech committed:
Update bbhw.fs
